src/tree/updater_quantile_hist.cc

```diff
@@ -732,198 +770,240 @@ void QuantileHistMaker::Builder::EvaluateSplit(const std::vector<ExpandEntry>& n
     }
   }
 
-  builder_monitor_.Stop("EvaluateSplit");
+  builder_monitor_.Stop("EvaluateSplits");
 }
 
-void QuantileHistMaker::Builder::ApplySplit(int nid,
-                                            const GHistIndexMatrix& gmat,
-                                            const ColumnMatrix& column_matrix,
-                                            const HistCollection& hist,
-                                            const DMatrix& fmat,
-                                            RegTree* p_tree) {
-  builder_monitor_.Start("ApplySplit");
-  // TODO(hcho3): support feature sampling by levels
-
-  /* 1. Create child nodes */
-  NodeEntry& e = snode_[nid];
-  bst_float left_leaf_weight =
-      spliteval_->ComputeWeight(nid, e.best.left_sum) * param_.learning_rate;
-  bst_float right_leaf_weight =
-      spliteval_->ComputeWeight(nid, e.best.right_sum) * param_.learning_rate;
-  p_tree->ExpandNode(nid, e.best.SplitIndex(), e.best.split_value,
-                     e.best.DefaultLeft(), e.weight, left_leaf_weight,
-                     right_leaf_weight, e.best.loss_chg, e.stats.sum_hess);
-
-  /* 2. Categorize member rows */
-  const auto nthread = static_cast<bst_omp_uint>(this->nthread_);
-  row_split_tloc_.resize(nthread);
-  for (bst_omp_uint i = 0; i < nthread; ++i) {
-    row_split_tloc_[i].left.clear();
-    row_split_tloc_[i].right.clear();
-  }
-  const bool default_left = (*p_tree)[nid].DefaultLeft();
-  const bst_uint fid = (*p_tree)[nid].SplitIndex();
-  const bst_float split_pt = (*p_tree)[nid].SplitCond();
-  const uint32_t lower_bound = gmat.cut.Ptrs()[fid];
-  const uint32_t upper_bound = gmat.cut.Ptrs()[fid + 1];
-  int32_t split_cond = -1;
-  // convert floating-point split_pt into corresponding bin_id
-  // split_cond = -1 indicates that split_pt is less than all known cut points
-  CHECK_LT(upper_bound,
-           static_cast<uint32_t>(std::numeric_limits<int32_t>::max()));
-  for (uint32_t i = lower_bound; i < upper_bound; ++i) {
-    if (split_pt == gmat.cut.Values()[i]) {
-      split_cond = static_cast<int32_t>(i);
+// split row indexes (rid_span) to 2 parts (left_part, right_part) depending
+// on comparison of indexes values (idx_span) and split point (split_cond)
+// Handle dense columns
+// Analog of std::stable_partition, but in no-inplace manner
+template <bool default_left>
+inline std::pair<size_t, size_t> PartitionDenseKernel(
+      common::Span<const size_t> rid_span, common::Span<const uint32_t> idx_span,
+      const int32_t split_cond, const uint32_t offset,
+      common::Span<size_t> left_part, common::Span<size_t> right_part) {
+  const uint32_t* idx = idx_span.data();
+  size_t* p_left_part = left_part.data();
+  size_t* p_right_part = right_part.data();
+  size_t nleft_elems = 0;
+  size_t nright_elems = 0;
+
+  const uint32_t missing_val = std::numeric_limits<uint32_t>::max();
+
+  for (auto rid : rid_span) {
```

Comment:
Just so you are aware, iterating through a Span implies boundary checks and may be slower than a standard for loop and accessing memory with pointers.

If this section is not performance critical you should absolutely iterate over span. I try to avoid optimising things that do not have visible effect on runtime.

Reply:
I tested performance again and I didn't observed regression vs. previous version.
So let's keep it as is (additional checks are not bad if it doesn't affect the performance).